laplace/baselaplace.py

```diff
@@ -1781,8 +1786,1062 @@ def sample(
+
+        """
+        if joint:
+            if self.diagonal_kernel:
+                kernel = torch.einsum("acp,bcp->abcc", jacobians, jacobians)
+            else:
+                kernel = torch.einsum("acp,bep->abce", jacobians, jacobians)
+
+        else:
+            if self.diagonal_kernel:
+                kernel = torch.empty(
+                    (jacobians.shape[0], self.n_outputs), device=jacobians.device
+                )
+                for c in range(self.n_outputs):
+                    kernel[:, c] = torch.norm(jacobians[:, c, :], dim=1) ** 2
+            else:
+                kernel = torch.einsum("bcp,bep->bce", jacobians, jacobians)
+        return kernel
+
+    def _kernel_batch_star(
+        self, jacobians: torch.Tensor, batch: torch.Tensor
+    ) -> torch.Tensor:
+        """Compute K_b_star, which is a part of K_M_star kernel matrix.
+
+        Parameters
+        ----------
+        jacobians : torch.Tensor (b1, C, P)
+        batch : torch.Tensor (b2, C)
+
+        Returns
+        -------
+        kernel : torch.tensor
+            K_batch_star with shape (b1, b2, C, C)
+        """
+        jacobians_2, _ = self._jacobians(batch)
+        if self.diagonal_kernel:
+            kernel = torch.empty(
+                (jacobians.shape[0], jacobians_2.shape[0], self.n_outputs),
+                device=jacobians.device,
+            )
+            for c in range(self.n_outputs):
+                kernel[:, :, c] = torch.einsum(
+                    "bp,ep->be", jacobians[:, c, :], jacobians_2[:, c, :]
+                )
+        else:
+            kernel = torch.einsum("bcp,dep->bdce", jacobians, jacobians_2)
+        return kernel
+
+    def _jacobians(self, X: torch.Tensor, enable_backprop: bool = None) -> tuple:
+        """A wrapper function to compute jacobians - this enables reusing same
+        kernel methods (kernel_batch etc.) in FunctionalLaplace and FunctionalLLLaplace
+        by simply overwriting this method instead of all kernel methods.
+        """
+        if enable_backprop is None:
+            enable_backprop = self.enable_backprop
+        return self.backend.jacobians(X, enable_backprop=enable_backprop)
+
+    def _mean_scatter_term_batch(
+        self, Js: torch.Tensor, f: torch.Tensor, y: torch.Tensor
+    ):
+        """Compute mean vector in the scatter term in the log marginal likelihood
+
+        See `scatter_lml` property above for the exact equations of mean vectors in scatter terms for
+        both types of likelihood (regression, classification).
+
+        Parameters
+        ----------
+        Js : torch.tensor
+              Jacobians (batch, output_shape, parameters)
+        f : torch.tensor
+              NN output (batch, output_shape)
+        y: torch.tensor
+              data labels (batch, output_shape)
+
+        Returns
+        -------
+        mu : torch.tensor
+            K_batch_star with shape (batch, output_shape)
+        """
+
```

Comment:
Not sure if a problem with Github's diff interface. But there is still an empty line here between the docstring and the function body. If this is true, please remove it.